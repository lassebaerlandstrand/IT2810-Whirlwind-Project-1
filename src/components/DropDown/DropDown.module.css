.container {
  position: relative;
  margin-block-start: 0;
  margin-block-end: 0;
  padding-inline-start: 0;
}

.dropDownButton {
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: fit-content;
  min-width: 10.5rem;
  padding: 0.5rem;
  background-color: white;
  border-radius: 0.3rem;
  border: 0.1rem var(--inactive-border) solid;
  cursor: pointer;
  box-sizing: border-box;
  user-select: none;
  font-size: 1rem;
}

.iconSpan {
  display: flex;
  align-items: center;
  justify-content: center;
  padding-left: 1rem;
  color: #999999;
  pointer-events: none;
}

.dropDownList {
  display: none;
  flex-direction: column;
  align-items: center;
  position: absolute;
  min-width: 100%;
  padding: 0;
  background-color: white;
  max-height: 20vh;
  overflow-y: scroll;
  scrollbar-width: none;
  -ms-overflow-style: none;
  box-sizing: border-box;
  list-style-type: none;
  user-select: none;
  margin-top: 0.25rem;
  border: 0.1rem var(--inactive-border) solid;
  border-radius: 0.3rem;
}

.dropDownList::-webkit-scrollbar {
  width: 0;
  height: 0;
}

.dropDownItem {
  width: 100%;
  padding: 0.5rem;
  cursor: pointer;
  box-sizing: border-box;
}

.dropDownItem:hover {
  background-color: #f4f4f4;
}

.buttonOpen {
  border: 0.1rem var(--active-border-blue) solid;
}

.listOpen {
  display: flex;
}

.label {
  font-size: 0.85rem;
}
