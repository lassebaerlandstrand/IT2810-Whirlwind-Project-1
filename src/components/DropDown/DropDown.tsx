import { IconChevronDown, IconChevronUp } from '@tabler/icons-react';
import { ButtonHTMLAttributes, useEffect, useId, useRef, useState } from 'react';
import styles from './DropDown.module.css';

type DropDownListProps = {
  open: boolean;
  options: string[];
  onSelected: (option: string) => void;
};

const DropDownList = ({ open, options, onSelected }: DropDownListProps) => {
  return (
    open && (
      <ul
        className={`${styles.dropDownList} ${open ? styles.listOpen : ''}`}
        aria-expanded={open}
        aria-labelledby="dropdown-button"
        role="listbox"
      >
        {options.map((option) => (
          <li
            key={option}
            className={styles.dropDownItem}
            onClick={() => onSelected(option)}
            role="option"
            aria-selected="false"
          >
            {option}
          </li>
        ))}
      </ul>
    )
  );
};

type DropDownButtonProps = {
  open: boolean;
  toggleOpen: () => void;
  selectedOption: string;
} & ButtonHTMLAttributes<HTMLButtonElement>;

const DropDownButton = ({ open, toggleOpen, selectedOption, ...rest }: DropDownButtonProps) => {
  return (
    <button
      className={`${styles.dropDownButton} ${open ? styles.buttonOpen : ''}`}
      onClick={toggleOpen}
      aria-expanded={open}
      aria-controls="dropdown-list"
      aria-haspopup="listbox"
      data-testid="dropdown-button"
      {...rest}
    >
      {selectedOption}{' '}
      <span className={styles.iconSpan}>
        {open ? <IconChevronUp aria-hidden="true" /> : <IconChevronDown aria-hidden="true" />}
      </span>
    </button>
  );
};

type DropDownProps = {
  options: string[];
  selectedOption: string;
  setSelectedOption: (option: string) => void;
  label: string;
};

const DropDown = ({ selectedOption, options, setSelectedOption, label }: DropDownProps) => {
  const [open, setOpen] = useState<boolean>(false);
  const dropDownRef = useRef<HTMLDivElement>(null);
  const dropDownId = useId();

  const toggleOpen = () => setOpen(!open);

  const handleSelectedOption = (option: string) => {
    setSelectedOption(option);
    setOpen(false);
  };

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (dropDownRef.current && !dropDownRef.current.contains(event.target as Node)) {
        setOpen(false);
      }
    };
    document.addEventListener('click', handleClickOutside);
    return () => document.removeEventListener('click', handleClickOutside);
  }, []);

  return (
    <menu className={styles.container} ref={dropDownRef}>
      <label className={styles.label} htmlFor={dropDownId}>
        {label}
      </label>
      <DropDownButton open={open} toggleOpen={toggleOpen} selectedOption={selectedOption} id={dropDownId} />
      <DropDownList
        open={open}
        options={options.filter((option) => option !== selectedOption)}
        onSelected={handleSelectedOption}
      />
    </menu>
  );
};

export default DropDown;
