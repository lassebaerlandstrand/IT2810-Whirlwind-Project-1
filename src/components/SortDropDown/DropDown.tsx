import React, { useEffect, useRef, useState } from 'react';
import styles from './DropDown.module.css';

type DropDownProps = {
  options: string[];
  selectedOption: string;
  setSelectedOption: (option: string) => void;
  label: string;
};

export const DropDown: React.FC<DropDownProps> = ({ selectedOption, options, setSelectedOption, label }) => {
  const [open, setOpen] = useState<boolean>(false);
  const dropDownRef = useRef<HTMLMenuElement>(null);

  const toggleOpen = () => setOpen((prev) => !prev);
  const handleOptionSelect = (option: string) => {
    setSelectedOption(option);
    setOpen(false);
  };

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (dropDownRef.current && !dropDownRef.current.contains(event.target as Node)) {
        setOpen(false);
      }
    };
    document.addEventListener('click', handleClickOutside);
    return () => document.removeEventListener('click', handleClickOutside);
  }, []);

  return (
    <div className={styles.container} ref={dropDownRef}>
      <label className={styles.label}>{label}</label>
      <button className={styles.dropDownButton} onClick={toggleOpen} aria-expanded={open} aria-haspopup="listbox">
        {selectedOption}
      </button>
      {open && (
        <ul className={styles.dropDownContent} role="listbox">
          {options.map((option) => (
            <li
              key={option}
              className={styles.dropDownItem}
              onClick={() => handleOptionSelect(option)}
              role="option"
              aria-selected={option === selectedOption}
            >
              {option}
            </li>
          ))}
        </ul>
      )}
    </div>
  );
};
