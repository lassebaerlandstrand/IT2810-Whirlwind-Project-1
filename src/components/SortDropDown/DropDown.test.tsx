import { fireEvent, render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import { DropDown } from './DropDown';

describe('DropDown Component', () => {
  const options = ['Option 1', 'Option 2', 'Option 3'];
  const label = 'Sort by:';

  it('should render with the correct default option', () => {
    const setSelectedOptionMock = vi.fn();
    render(
      <DropDown selectedOption="Option 1" options={options} setSelectedOption={setSelectedOptionMock} label={label} />,
    );

    expect(screen.getByText('Option 1')).toBeInTheDocument();
    expect(screen.getByText('Sort by:')).toBeInTheDocument();
  });

  it('should open and close the dropdown when clicked', () => {
    const setSelectedOptionMock = vi.fn();
    render(
      <DropDown selectedOption="Option 1" options={options} setSelectedOption={setSelectedOptionMock} label={label} />,
    );

    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(screen.getByRole('listbox')).toBeInTheDocument();

    fireEvent.click(button);
    expect(screen.queryByRole('listbox')).not.toBeInTheDocument();
  });

  it('should change the selected option when an item is clicked', () => {
    const setSelectedOptionMock = vi.fn();
    render(
      <DropDown selectedOption="Option 1" options={options} setSelectedOption={setSelectedOptionMock} label={label} />,
    );

    fireEvent.click(screen.getByRole('button'));
    fireEvent.click(screen.getByText('Option 2'));

    expect(setSelectedOptionMock).toHaveBeenCalledWith('Option 2');
  });

  it('should close the dropdown when clicking outside of it', () => {
    const setSelectedOptionMock = vi.fn();
    render(
      <DropDown selectedOption="Option 1" options={options} setSelectedOption={setSelectedOptionMock} label={label} />,
    );

    fireEvent.click(screen.getByRole('button'));
    expect(screen.getByRole('listbox')).toBeInTheDocument();

    fireEvent.click(document.body);
    expect(screen.queryByRole('listbox')).not.toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const { asFragment } = render(
      <DropDown selectedOption="Option 1" options={options} setSelectedOption={vi.fn()} label={label} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
