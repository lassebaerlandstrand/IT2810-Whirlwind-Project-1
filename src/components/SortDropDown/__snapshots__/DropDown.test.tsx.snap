// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DropDown Component > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="_container_bfe205"
  >
    <label
      class="_label_bfe205"
    >
      Sort by:
    </label>
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      class="_dropDownButton_bfe205"
    >
      Option 1
      <span
        class="_iconSpan_bfe205"
      >
        â–¼
      </span>
    </button>
  </div>
</DocumentFragment>
`;
